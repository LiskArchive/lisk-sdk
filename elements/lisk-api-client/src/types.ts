/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { Schema } from '@liskhq/lisk-codec';

export interface Defer<T> {
	promise: Promise<T>;
	resolve: (result: T) => void;
	reject: (error?: Error) => void;
}

export interface JSONRPCNotification<T> {
	readonly id: never;
	readonly jsonrpc: string;
	readonly method: string;
	readonly params?: T;
}

export interface JSONRPCError {
	code: number;
	message: string;
	data?: string | number | boolean | Record<string, unknown>;
}

export interface JSONRPCResponse<T> {
	readonly id: number;
	readonly jsonrpc: string;
	readonly method: never;
	readonly params: never;
	readonly error?: JSONRPCError;
	readonly result?: T;
}

export type JSONRPCMessage<T> = JSONRPCNotification<T> | JSONRPCResponse<T>;

export type EventCallback<T = Record<string, unknown>> = (event?: T) => void | Promise<void>;

export interface Channel {
	connect: () => Promise<void>;
	disconnect: () => Promise<void>;
	invoke: <T>(actionName: string, params?: Record<string, unknown>) => Promise<T>;
	subscribe: (eventName: string, cb: EventCallback) => void;
}

export interface RegisteredSchemas {
	block: Schema;
	header: Schema;
	asset: Schema;
	transaction: Schema;
	event: Schema;
}

export interface ModuleMetadata {
	id: string;
	name: string;
	endpoints: {
		name: string;
		request?: Schema;
		response: Schema;
	}[];
	events: {
		typeID: string;
		data: Schema;
	}[];
	commands: {
		id: string;
		name: string;
		params?: Schema;
	}[];
	assets: {
		version: number;
		data: Schema;
	}[];
}

export interface GenesisConfig {
	[key: string]: unknown;
	readonly bftThreshold: number;
	readonly communityIdentifier: string;
	readonly blockTime: number;
	readonly maxTransactionsSize: number;
	readonly rewards: {
		readonly milestones: string[];
		readonly offset: number;
		readonly distance: number;
	};
	readonly minFeePerByte: number;
}

export interface NodeInfo {
	readonly version: string;
	readonly networkVersion: string;
	readonly networkIdentifier: string;
	readonly lastBlockID: string;
	readonly height: number;
	readonly genesisHeight: number;
	readonly finalizedHeight: number;
	readonly syncing: boolean;
	readonly unconfirmedTransactions: number;
	readonly genesisConfig: GenesisConfig;
	readonly network: {
		readonly port: number;
		readonly hostIp?: string;
		readonly seedPeers: {
			readonly ip: string;
			readonly port: number;
		}[];
		readonly blacklistedIPs?: string[];
		readonly fixedPeers?: string[];
		readonly whitelistedPeers?: {
			readonly ip: string;
			readonly port: number;
		}[];
	};
}

export interface MultiSignatureKeys {
	readonly mandatoryKeys: Buffer[];
	readonly optionalKeys: Buffer[];
	readonly numberOfSignatures: number;
}

export interface NetworkStats {
	[key: string]: unknown;
	readonly outgoing: {
		count: number;
		connects: number;
		disconnects: number;
	};
	readonly incoming: {
		count: number;
		connects: number;
		disconnects: number;
	};
	readonly banning: {
		count: number;
		bannedPeers: {
			[key: string]: {
				lastBanTime: number;
				banCount: number;
			};
		};
	};
	readonly totalConnectedPeers: number;
	readonly totalDisconnectedPeers: number;
	readonly totalErrors: number;
	readonly totalRemovedPeers: number;
	readonly totalMessagesReceived: {
		[key: string]: number;
	};
	readonly totalRequestsReceived: {
		[key: string]: number;
	};
	readonly totalPeersDiscovered: number;
	readonly startTime: number;
}

export interface PeerInfo {
	readonly ipAddress: string;
	readonly port: number;
	readonly networkIdentifier?: string;
	readonly networkVersion?: string;
	readonly nonce?: string;
	readonly options?: { [key: string]: unknown };
}

type Primitive = string | number | bigint | boolean | null | undefined;
type Replaced<T, TReplace, TWith, TKeep = Primitive> = T extends TReplace | TKeep
	? T extends TReplace
		? TWith | Exclude<T, TReplace>
		: T
	: {
			[P in keyof T]: Replaced<T[P], TReplace, TWith, TKeep>;
	  };

export type JSONObject<T> = Replaced<T, bigint | Buffer, string>;

export interface BlockHeader {
	readonly version: number;
	readonly height: number;
	readonly generatorAddress: Buffer;
	readonly previousBlockID: Buffer;
	readonly timestamp: number;
	readonly maxHeightPrevoted: number;
	readonly maxHeightGenerated: number;
	readonly aggregateCommit: {
		readonly height: number;
		readonly aggregationBits: Buffer;
		readonly certificateSignature: Buffer;
	};
	readonly validatorsHash: Buffer;
	readonly stateRoot: Buffer;
	readonly transactionRoot: Buffer;
	readonly assetRoot: Buffer;
	readonly eventRoot: Buffer;
	readonly signature: Buffer;
	readonly id: Buffer;
}

export type BlockHeaderJSON = JSONObject<BlockHeader>;

export interface BlockAsset {
	module: Buffer;
	data: Buffer;
}

export type BlockAssetJSON = JSONObject<BlockAsset>;
export type DecodedBlockAsset<T = Record<string, unknown>> = Omit<BlockAsset, 'data'> & { data: T };
export type DecodedBlockAssetJSON<T = Record<string, unknown>> = Omit<BlockAssetJSON, 'data'> & {
	data: T;
};

export interface Transaction {
	readonly module: string;
	readonly command: string;
	readonly senderPublicKey: Buffer;
	readonly nonce: bigint;
	readonly fee: bigint;
	readonly params: Buffer;
	readonly signatures: ReadonlyArray<Buffer>;
	readonly id: Buffer;
}

export type TransactionJSON = JSONObject<Transaction>;
export type DecodedTransaction<T = Record<string, unknown>> = Omit<Transaction, 'params'> & {
	params: T;
};
export type DecodedTransactionJSON<T = Record<string, unknown>> = Omit<
	TransactionJSON,
	'params'
> & { params: T };

export interface Block {
	header: BlockHeader;
	transactions: Transaction[];
	assets: BlockAsset[];
}

export interface DecodedBlock {
	header: BlockHeader;
	transactions: DecodedTransaction[];
	assets: DecodedBlockAsset[];
}

export interface BlockJSON {
	header: BlockHeaderJSON;
	transactions: TransactionJSON[];
	assets: BlockAssetJSON[];
}

export interface DecodedBlockJSON {
	header: BlockHeaderJSON;
	transactions: DecodedTransactionJSON[];
	assets: DecodedBlockAssetJSON[];
}

export interface Event {
	readonly moduleID: Buffer;
	readonly typeID: Buffer;
	readonly topics: Buffer[];
	readonly index: number;
	readonly data: Buffer;
}

export type EventJSON = JSONObject<Event>;
export type DecodedEvent<T = Record<string, unknown>> = Omit<Event, 'data'> & { data: T };
