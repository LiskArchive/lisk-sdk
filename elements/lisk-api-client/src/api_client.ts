/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import * as os from 'os';
import { HashMap, InitOptions } from './api_types';
import * as constants from './constants';
import { AccountsResource } from './resources/accounts';
import { BlocksResource } from './resources/blocks';
import { DappsResource } from './resources/dapps';
import { DelegatesResource } from './resources/delegates';
import { NodeResource } from './resources/node';
import { PeersResource } from './resources/peers';
import { SignaturesResource } from './resources/signatures';
import { TransactionsResource } from './resources/transactions';
import { VotersResource } from './resources/voters';
import { VotesResource } from './resources/votes';

const defaultOptions = {
	bannedNodes: [],
	randomizeNodes: true,
};

const commonHeaders: HashMap = {
	Accept: 'application/json',
	'Content-Type': 'application/json',
};

const getClientHeaders = (clientOptions: ClientOptions): HashMap => {
	const { name = '????', version = '????', engine = '????' } = clientOptions;

	const liskElementsInformation =
		'LiskElements/1.0 (+https://github.com/LiskHQ/lisk-elements)';
	const locale: string | undefined =
		process.env.LC_ALL ||
		process.env.LC_MESSAGES ||
		process.env.LANG ||
		process.env.LANGUAGE;
	const systemInformation = `${os.platform()} ${os.release()}; ${os.arch()}${
		locale ? `; ${locale}` : ''
	}`;

	return {
		'User-Agent': `${name}/${version} (${engine}) ${liskElementsInformation} ${systemInformation}`,
	};
};

export interface ClientOptions {
	readonly engine?: string;
	readonly name?: string;
	readonly version?: string;
}

export class APIClient {
	public static get constants(): typeof constants {
		return constants;
	}

	public static createMainnetAPIClient(options?: InitOptions): APIClient {
		return new APIClient(constants.MAINNET_NODES, {
			nethash: constants.MAINNET_NETHASH,
			...options,
		});
	}

	public static createTestnetAPIClient(options?: InitOptions): APIClient {
		return new APIClient(constants.TESTNET_NODES, {
			nethash: constants.TESTNET_NETHASH,
			...options,
		});
	}

	public accounts: AccountsResource;
	public bannedNodes!: ReadonlyArray<string>;
	public blocks: BlocksResource;
	public currentNode!: string;
	public dapps: DappsResource;
	public delegates: DelegatesResource;
	public headers!: HashMap;
	public node: NodeResource;
	public nodes!: ReadonlyArray<string>;
	public peers: PeersResource;
	public randomizeNodes!: boolean;
	public signatures: SignaturesResource;
	public transactions: TransactionsResource;
	public voters: VotersResource;
	public votes: VotesResource;

	public constructor(
		nodes: ReadonlyArray<string>,
		providedOptions: InitOptions = {},
	) {
		this.initialize(nodes, providedOptions);
		this.accounts = new AccountsResource(this);
		this.blocks = new BlocksResource(this);
		this.dapps = new DappsResource(this);
		this.delegates = new DelegatesResource(this);
		this.node = new NodeResource(this);
		this.peers = new PeersResource(this);
		this.signatures = new SignaturesResource(this);
		this.transactions = new TransactionsResource(this);
		this.voters = new VotersResource(this);
		this.votes = new VotesResource(this);
	}

	public banActiveNode(): boolean {
		return this.banNode(this.currentNode);
	}

	public banActiveNodeAndSelect(): boolean {
		const banned = this.banActiveNode();
		if (banned) {
			this.currentNode = this.getNewNode();
		}

		return banned;
	}

	public banNode(node: string): boolean {
		if (!this.isBanned(node)) {
			this.bannedNodes = [...this.bannedNodes, node];

			return true;
		}

		return false;
	}

	public getNewNode(): string {
		const nodes = this.nodes.filter(
			(node: string): boolean => !this.isBanned(node),
		);

		if (nodes.length === 0) {
			throw new Error('Cannot get new node: all nodes have been banned.');
		}

		const randomIndex = Math.floor(Math.random() * nodes.length);

		return nodes[randomIndex];
	}

	public hasAvailableNodes(): boolean {
		return this.nodes.some((node: string): boolean => !this.isBanned(node));
	}

	public initialize(
		nodes: ReadonlyArray<string>,
		providedOptions: InitOptions = {},
	): void {
		if (!Array.isArray(nodes) || nodes.length <= 0) {
			throw new Error('APIClient requires nodes for initialization.');
		}

		if (typeof providedOptions !== 'object' || Array.isArray(providedOptions)) {
			throw new Error(
				'APIClient takes an optional object as the second parameter.',
			);
		}

		const options: InitOptions = { ...defaultOptions, ...providedOptions };

		this.headers = {
			...commonHeaders,
			...(options.nethash ? { nethash: options.nethash } : {}),
			...(options.client ? getClientHeaders(options.client) : {}),
		};

		this.nodes = nodes;
		this.bannedNodes = [...(options.bannedNodes || [])];
		this.currentNode = options.node || this.getNewNode();
		this.randomizeNodes = options.randomizeNodes !== false;
	}

	public isBanned(node: string): boolean {
		return this.bannedNodes.includes(node);
	}
}
