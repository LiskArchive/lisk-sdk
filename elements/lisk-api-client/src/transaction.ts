/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import {
	signTransaction,
	signMultiSignatureTransaction,
	computeMinFee,
} from '@liskhq/lisk-transactions';
import {
	getAddressAndPublicKeyFromPassphrase,
	getAddressFromPublicKey,
} from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import {
	decodeTransaction,
	encodeTransaction,
	getTransactionAssetSchema,
	decodeAccount,
} from './codec';
import { Channel, RegisteredSchemas, NodeInfo } from './types';

interface MultiSignatureKeys {
	readonly mandatoryKeys: Buffer[];
	readonly optionalKeys: Buffer[];
	readonly numberOfSignatures: number;
}

export class Transaction {
	private readonly _channel: Channel;
	private readonly _schema: RegisteredSchemas;
	private readonly _nodeInfo: NodeInfo;

	public constructor(channel: Channel, registeredSchema: RegisteredSchemas, nodeInfo: NodeInfo) {
		this._channel = channel;
		this._schema = registeredSchema;
		this._nodeInfo = nodeInfo;
	}

	public async create(
		input: {
			moduleID?: number; // id takes priority
			moduleName?: string;
			assetID?: number; // id takes priority
			assetName?: string;
			fee: bigint;
			nonce?: bigint;
			senderPublicKey?: Buffer;
			asset: Record<string, unknown>;
			signatures?: Buffer[];
		},
		passphrase: string,
		options?: {
			includeSenderSignature?: boolean;
			multisignatureKeys?: {
				mandatoryKeys: Buffer[];
				optionalKeys: Buffer[];
			};
		},
	): Promise<Record<string, unknown>> {
		const txInput = input;
		const networkIdentifier = Buffer.from(this._nodeInfo.networkIdentifier, 'hex');
		const { publicKey, address } = getAddressAndPublicKeyFromPassphrase(passphrase);
		const accountHex = await this._channel.invoke<string>('app:getAccount', {
			address: address.toString('hex'),
		});
		const account = decodeAccount(Buffer.from(accountHex, 'hex'), this._schema);
		if (!txInput.moduleID) {
			if (!txInput.moduleName) {
				throw new Error('Missing moduleID and moduleName');
			}
			const registeredModule = this._nodeInfo.registeredModules.find(
				module => module.name === input.moduleName,
			);
			if (!registeredModule) {
				throw new Error(`Module corresponding to name ${txInput.moduleName} not registered.`);
			}
			txInput.moduleID = registeredModule.id;
		}
		if (typeof txInput.assetID !== 'number') {
			if (!txInput.assetName) {
				throw new Error('Missing assetID and assetName');
			}
			const registeredModule = this._nodeInfo.registeredModules.find(
				m => m.id === txInput.moduleID,
			);
			if (!registeredModule) {
				throw new Error(`Module corresponding to id ${txInput.moduleID} not registered.`);
			}
			const registeredAsset = registeredModule.transactionAssets.find(
				asset => asset.name === txInput.assetName,
			);
			if (!registeredAsset) {
				throw new Error(`Asset corresponding to name ${txInput.assetName} not registered.`);
			}
			txInput.assetID = registeredAsset.id;
		}
		if (typeof txInput.nonce !== 'bigint') {
			if (
				typeof account.sequence !== 'object' ||
				typeof (account.sequence as Record<string, unknown>).nonce !== 'bigint'
			) {
				throw new Error('Unsupported account type');
			}
			txInput.nonce = (account.sequence as { nonce: bigint }).nonce;
		}
		if (txInput.nonce < BigInt(0)) {
			throw new Error('Nonce must be greater or equal to zero');
		}
		if (!txInput.senderPublicKey) {
			txInput.senderPublicKey = publicKey;
		}
		// If signature is not set, assign empty array
		txInput.signatures = txInput.signatures ?? [];
		const assetSchema = getTransactionAssetSchema(txInput, this._schema);
		if (account.keys && (account.keys as MultiSignatureKeys).numberOfSignatures > 0) {
			return signMultiSignatureTransaction(
				assetSchema,
				txInput,
				networkIdentifier,
				passphrase,
				account.keys as MultiSignatureKeys,
				options?.includeSenderSignature,
			);
		}
		if (options?.multisignatureKeys && options?.includeSenderSignature) {
			return signMultiSignatureTransaction(
				assetSchema,
				txInput,
				networkIdentifier,
				passphrase,
				options.multisignatureKeys,
				options.includeSenderSignature,
			);
		}
		return signTransaction(assetSchema, txInput, networkIdentifier, passphrase);
	}

	public async get(id: Buffer): Promise<Record<string, unknown>> {
		const transactionHex = await this._channel.invoke<string>('app:getTransactionByID', {
			id: id.toString('hex'),
		});
		return decodeTransaction(Buffer.from(transactionHex, 'hex'), this._schema);
	}

	public async getFromPool(): Promise<Record<string, unknown>[]> {
		const transactionsHex = await this._channel.invoke<string[]>('app:getTransactionsFromPool');
		const decodedTransactions: Record<string, unknown>[] = [];
		for (const transactionHex of transactionsHex) {
			decodedTransactions.push(decodeTransaction(Buffer.from(transactionHex, 'hex'), this._schema));
		}
		return decodedTransactions;
	}

	public async sign(
		transaction: Record<string, unknown>,
		passphrases: string[],
		options?: {
			includeSenderSignature?: boolean;
			multisignatureKeys?: {
				mandatoryKeys: Buffer[];
				optionalKeys: Buffer[];
			};
		},
	): Promise<Record<string, unknown>> {
		const assetSchema = getTransactionAssetSchema(transaction, this._schema);
		const networkIdentifier = Buffer.from(this._nodeInfo.networkIdentifier, 'hex');
		const address = getAddressFromPublicKey(transaction.senderPublicKey as Buffer);
		const accountHex = await this._channel.invoke<string>('app:getAccount', {
			address: address.toString('hex'),
		});
		const account = decodeAccount(Buffer.from(accountHex, 'hex'), this._schema);
		if (account.keys && (account.keys as MultiSignatureKeys).numberOfSignatures > 0) {
			for (const passphrase of passphrases) {
				signMultiSignatureTransaction(
					assetSchema,
					transaction,
					networkIdentifier,
					passphrase,
					account.keys as MultiSignatureKeys,
					options?.includeSenderSignature,
				);
			}
			return transaction;
		}
		if (options?.multisignatureKeys && options?.includeSenderSignature) {
			for (const passphrase of passphrases) {
				signMultiSignatureTransaction(
					assetSchema,
					transaction,
					networkIdentifier,
					passphrase,
					options.multisignatureKeys as MultiSignatureKeys,
					options.includeSenderSignature,
				);
			}
			return transaction;
		}
		return signTransaction(assetSchema, transaction, networkIdentifier, passphrases[0]);
	}

	public async send(
		transaction: Record<string, unknown>,
	): Promise<{
		transactionId: string;
	}> {
		const encodedTx = encodeTransaction(transaction, this._schema);
		return this._channel.invoke<{
			transactionId: string;
		}>('app:postTransaction', { transaction: encodedTx.toString('hex') });
	}

	public decode<T = Record<string, unknown>>(transaction: Buffer): T {
		return decodeTransaction(transaction, this._schema) as T;
	}

	public encode(transaction: Record<string, unknown>): Buffer {
		return encodeTransaction(transaction, this._schema);
	}

	public computeMinFee(transaction: Record<string, unknown>): bigint {
		const assetSchema = getTransactionAssetSchema(transaction, this._schema);
		return computeMinFee(assetSchema, transaction);
	}

	public toJSON(transaction: Record<string, unknown>): Record<string, unknown> {
		const { asset: txAsset, ...txRoot } = transaction;
		// We need to do this as our schemas do not include the ID. Keep this.
		const tmpId = txRoot.id;
		delete txRoot.id;

		const schemaAsset = getTransactionAssetSchema(txRoot, this._schema);
		const jsonTxAsset = codec.toJSON(schemaAsset, txAsset as Record<string, unknown>);
		const jsonTxRoot = codec.toJSON(this._schema.transaction, txRoot);

		const jsonTx = {
			...jsonTxRoot,
			asset: jsonTxAsset,
			id: Buffer.isBuffer(tmpId) ? tmpId.toString('hex') : tmpId,
		};

		return jsonTx;
	}

	public fromJSON(transaction: Record<string, unknown>): Record<string, unknown> {
		const { asset: txAsset, ...txRoot } = transaction;
		// We need to do this as our schemas do not include the ID. Keep this.
		const tmpId = txRoot.id;
		delete txRoot.id;

		const schemaAsset = getTransactionAssetSchema(txRoot, this._schema);
		const txAssetObject = codec.fromJSON(schemaAsset, txAsset as Record<string, unknown>);
		const txRootObject = codec.fromJSON(this._schema.transaction, txRoot);

		const txObject = {
			...txRootObject,
			asset: txAssetObject,
			id: typeof tmpId === 'string' ? Buffer.from(tmpId, 'hex') : Buffer.alloc(0),
		};

		return txObject;
	}
}
