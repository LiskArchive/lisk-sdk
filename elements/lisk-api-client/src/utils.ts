/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { HashMap } from './api_types';

export const toQueryString = (obj: HashMap): string => {
	const parts = Object.keys(obj).reduce(
		(
			accumulator: ReadonlyArray<string>,
			key: string,
		): ReadonlyArray<string> => [
			...accumulator,
			`${encodeURIComponent(key)}=${encodeURIComponent(obj[key])}`,
		],
		[],
	);

	return parts.join('&');
};

const urlParamRegex = /{[^}]+}/;
export const solveURLParams = (url: string, params: HashMap = {}): string => {
	if (Object.keys(params).length === 0) {
		if (url.match(urlParamRegex) !== null) {
			throw new Error('URL is not completely solved');
		}

		return url;
	}
	const solvedURL = Object.keys(params).reduce(
		(accumulator: string, key: string): string =>
			accumulator.replace(`{${key}}`, params[key]),
		url,
	);

	if (solvedURL.match(urlParamRegex) !== null) {
		throw new Error('URL is not completely solved');
	}

	return encodeURI(solvedURL);
};
