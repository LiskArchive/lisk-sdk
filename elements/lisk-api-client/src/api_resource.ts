/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import Axios, { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
import { APIClient } from './api_client';
import { APIErrorResponse, APIResponse, HashMap } from './api_types';
import { APIError } from './errors';

const API_RECONNECT_MAX_RETRY_COUNT = 3;

const REQUEST_RETRY_TIMEOUT = 1000;

export class APIResource {
	public apiClient: APIClient;
	public path: string;

	public constructor(apiClient: APIClient) {
		this.apiClient = apiClient;
		this.path = '';
	}

	public get headers(): HashMap {
		return this.apiClient.headers;
	}

	public get resourcePath(): string {
		return `${this.apiClient.currentNode}/api${this.path}`;
	}

	public async handleRetry(
		error: Error,
		req: AxiosRequestConfig,
		retryCount: number,
	): Promise<APIResponse> {
		if (this.apiClient.hasAvailableNodes()) {
			return new Promise<APIResponse>(resolve =>
				setTimeout(resolve, REQUEST_RETRY_TIMEOUT),
			).then(
				async (): Promise<APIResponse> => {
					if (retryCount > API_RECONNECT_MAX_RETRY_COUNT) {
						throw error;
					}
					if (this.apiClient.randomizeNodes) {
						this.apiClient.banActiveNodeAndSelect();
					}

					return this.request(req, true, retryCount + 1);
				},
			);
		}

		return Promise.reject(error);
	}

	public async request(
		req: AxiosRequestConfig,
		retry: boolean,
		retryCount: number = 1,
	): Promise<APIResponse> {
		const request = Axios.request(req)
			.then((res: AxiosResponse) => res.data)
			.catch(
				(error: AxiosError): void => {
					if (error.response) {
						const { status } = error.response;
						if (error.response.data) {
							const {
								error: errorString,
								errors,
								message,
							}: APIErrorResponse = error.response.data;
							throw new APIError(
								message || errorString || 'An unknown error has occurred.',
								status,
								errors,
							);
						}
						throw new APIError('An unknown error has occurred.', status);
					}
					throw error;
				},
			);

		if (retry) {
			return request.catch(async (err: Error) =>
				this.handleRetry(err, req, retryCount),
			);
		}

		return request;
	}
}
