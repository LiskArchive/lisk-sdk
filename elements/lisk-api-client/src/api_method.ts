/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { AxiosRequestConfig } from 'axios';
import {
	APIHandler,
	APIResponse,
	HashMap,
	RequestConfig,
	Resource,
} from './api_types';
import { GET } from './constants';
import { solveURLParams, toQueryString } from './utils';

// Bind to resource class
export const apiMethod = (options: RequestConfig = {}): APIHandler =>
	async function apiHandler(
		this: Resource,
		// tslint:disable-next-line readonly-array
		...args: Array<number | string | object>
	): Promise<APIResponse> {
		const {
			method = GET,
			path = '',
			urlParams = [],
			validator,
			defaultData = {},
			retry = false,
		} = options;

		if (urlParams.length > 0 && args.length < urlParams.length) {
			return Promise.reject(
				new Error(
					`This endpoint must be supplied with the following parameters: ${urlParams.toString()}`,
				),
			);
		}

		const data = {
			...defaultData,
			...(args.length > urlParams.length &&
			typeof args[urlParams.length] === 'object'
				? (args[urlParams.length] as object)
				: {}),
		};

		if (validator) {
			try {
				validator(data);
			} catch (err) {
				return Promise.reject(err);
			}
		}

		const resolvedURLObject = urlParams.reduce(
			// tslint:disable-next-line no-inferred-empty-object-type
			(accumulator: HashMap, param: string, i: number): HashMap => {
				const value = args[i];
				if (typeof value !== 'string' && typeof value !== 'number') {
					throw new Error('Parameter must be a string or a number');
				}

				return {
					...accumulator,
					[param]: typeof value === 'number' ? value.toString() : value,
				};
			},
			{},
		);

		const requestData: AxiosRequestConfig = {
			headers: this.headers,
			method,
			url: solveURLParams(`${this.resourcePath}${path}`, resolvedURLObject),
		};

		if (Object.keys(data).length > 0) {
			if (method === GET) {
				requestData.url += `?${toQueryString(data)}`;
			} else {
				requestData.data = data;
			}
		}

		return this.request(requestData, retry);
	};
