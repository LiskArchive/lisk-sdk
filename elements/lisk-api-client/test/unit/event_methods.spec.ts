/*
 * Copyright Â© 2023 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { Channel } from '../../src/types';
import { EventMethods } from '../../src/event_methods';
import { metadata } from '../utils/transaction';
import { events as encodedEventsJSON } from '../fixtures/encoded_events.json';

describe('event', () => {
	const BLOCK_HEIGHT = encodedEventsJSON[0].height;
	let channel: Channel;
	let event: EventMethods;

	beforeEach(() => {
		channel = {
			connect: jest.fn(),
			disconnect: jest.fn(),
			invoke: jest.fn().mockResolvedValue(encodedEventsJSON),
			// eslint-disable-next-line @typescript-eslint/no-misused-promises
			subscribe: jest.fn((_, callback) => callback({ blockHeader: { BLOCK_HEIGHT } })),
		};
		event = new EventMethods(channel, metadata);
	});

	describe('get', () => {
		it('should decode all events', async () => {
			expect(await event.get(BLOCK_HEIGHT)).toMatchSnapshot();
		});

		it('should return only decoded events that match specified module name', async () => {
			const decodedEvents = await event.get(BLOCK_HEIGHT, { module: 'pos' });

			expect(decodedEvents).toHaveLength(2);
			expect(decodedEvents[0].module).toBe('pos');
			expect(decodedEvents[0].name).toBe('validatorStaked');
			expect(decodedEvents[1].module).toBe('pos');
			expect(decodedEvents[1].name).toBe('commandExecutionResult');
		});

		it('should return only decoded events that match specified module and event name', async () => {
			const decodedEvents = await event.get(BLOCK_HEIGHT, {
				module: 'pos',
				name: 'validatorStaked',
			});

			expect(decodedEvents).toHaveLength(1);
			expect(decodedEvents[0].module).toBe('pos');
			expect(decodedEvents[0].name).toBe('validatorStaked');
		});
	});

	describe('subscribe', () => {
		it('should subscribe to all events', async () => {
			const decodedEvents = await event.get(BLOCK_HEIGHT);

			event.subscribe(events => {
				expect(events).toEqual(decodedEvents);
			});

			expect.assertions(1);
		});
	});
});
