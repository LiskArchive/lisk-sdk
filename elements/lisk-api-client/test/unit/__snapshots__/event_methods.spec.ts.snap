// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event get should decode all events 1`] = `
[
  {
    "data": {
      "address": "lske5sqed53fdcs4m9et28f2k7u9fk6hno9bauday",
      "amount": "10000000",
      "module": "fee",
      "result": 0,
      "tokenID": "0400000000000000",
    },
    "height": 1345,
    "index": 0,
    "module": "token",
    "name": "lock",
    "topics": [
      "0b3133525e7aeeffeea6c84375f029dfd7dd68c4ade9b253bd8f024829865f73",
      "b2f75b6968f687b495b69e1fec51d06fa8ae9d8d",
    ],
  },
  {
    "data": {
      "amount": "5000000000",
      "result": 1,
      "senderAddress": "lske5sqed53fdcs4m9et28f2k7u9fk6hno9bauday",
      "validatorAddress": "lsk8z34uaatynq6nh2dqasqpfsn8t9saxn6dpa536",
    },
    "height": 1345,
    "index": 1,
    "module": "pos",
    "name": "validatorStaked",
    "topics": [
      "0b3133525e7aeeffeea6c84375f029dfd7dd68c4ade9b253bd8f024829865f73",
      "b2f75b6968f687b495b69e1fec51d06fa8ae9d8d",
      "781098673389eaa3f71aae77527b677cdbbc84ea",
    ],
  },
  {
    "data": {
      "success": false,
    },
    "height": 1345,
    "index": 6,
    "module": "pos",
    "name": "commandExecutionResult",
    "topics": [
      "0b3133525e7aeeffeea6c84375f029dfd7dd68c4ade9b253bd8f024829865f73",
    ],
  },
  {
    "data": {
      "amount": "0",
      "reduction": 1,
    },
    "height": 1345,
    "index": 7,
    "module": "dynamicReward",
    "name": "rewardMinted",
    "topics": [
      "03",
      "906f6b4539a3d16a215f629946064a3065626848",
    ],
  },
]
`;
